index.html

```diff
@@ -1690,59 +1690,93 @@ <h3>Generate Proof</h3>
+          </li>
+          <li>
+If the cryptographic suite does not support unlinkability and the
+<var>proof</var>.<var>created</var> value is not set, a `MALFORMED_PROOF_ERROR`
+MUST be raised.
+          </li>
+          <li>
+Let <var>unsecuredDocument</var> be a copy of <var>securedDocument</var> with
+the `proof` value removed.
+          </li>
+          <li>
+Let <var>transformedDocument</var> be the result of
+<a href="#dfn-transformation">transforming</a> <var>unsecuredDocument</var>
+according to a <a>transformation algorithm</a> associated with the
+<var>cryptosuite</var> specified in <var>proof</var> and the <var>options</var>
+parameters provided as inputs to the algorithm.
+        </li>
+          <li>
+Let <var>hashData</var> be the result of
+<a href="#dfn-hashing">hashing</a> <var>transformedDocument</var>
+according to a <a>hashing algorithm</a> associated with the
+<var>cryptosuite</var> specified in <var>proof</var> and the <var>options</var>
+parameters provided as inputs to the algorithm.
           </li>
           <li>
-Let <var>document</var> be a copy of <var>signed document</var>.
+Let <var>isProofValid</var> be the result of running the
+<a>proof verification algorithm</a> associated with the <var>cryptosuite</var>
+with the <var>hashData</var> and <var>options</var> parameters provided
+as inputs to the algorithm.
+          </li>
+          <li>
+If the <var>proof</var>.<var>proofPurpose</var> value does not match
+<var>options</var>.<var>proofPurpose</var>, a `MISMATCHED_PROOF_PURPOSE_ERROR`
+MUST be raised.
           </li>
           <li>
-Remove any <code>proof</code> nodes from the default graph in
-<var>document</var> and save it as <var>proof</var>.
+If the <var>proof</var>.<var>created</var> is set and it deviates more than
+<var>options</var>.<var>acceptableCreatedTimeDeviationInSeconds</var> seconds, a
+`CREATED_TIME_DEVIATION_ERROR` MUST be raised.
           </li>
           <li>
-Generate a <var>transformed document</var> by canonicalizing
-<var>document</var> according to the <a>transformation algorithm</a>
-(e.g. the </var><em>URDNA2015</em> [[!RDF-DATASET-C14N]] algorithm).
+If <var>options</var>.<var>domain</var> is set and it does not match
+<var>proof</var>.<var>domain</var>, an `INVALID_DOMAIN_ERROR` MUST be raised.
           </li>
           <li>
-Create a value <var>tbv</var> that represents the data to be verified, and
-set it to the result of running the
-<a href="#dfn-hashing-algorithm">hashing algorithm</a>,
-passing the information in <var>proof</var>.
+If <var>options</var>.<var>challenge</var> is set and it does not match
+<var>proof</var>.<var>challenge</var>, an `INVALID_CHALLENGE_ERROR` MUST be
+raised.
           </li>
           <li>
-Pass the <a>proofValue</a>, <var>tbv</var>,
-and the <a>public key</a> to the <a>proof verification algorithm</a>.
-Return the resulting boolean value.
+Return <var>isProofValid</var> as the <a>verification result</a>.
```

Comment:
```suggestion
Return <var>isProofVerified</var> as the <a>verification result</a>.
```

Reply:
I think we need a cleaner separation of cryptosuite verification and purpose verification. We should:

1. Find a proof that matches the expected purpose; if none is found, raise an error.
2. Verify the proof using the cryptosuite. This returns `true` or `false` or throws if the operation cannot be executed.
3. If cryptosuite verification failed, return that the proof is not verified, otherwise...
4. Verify the proof against the expected purpose's specific checks (checking 'domain' and 'challenge' and so on are run here for authentication proofs, allowing for other purposes to be slotted in, i.e., "authentication" isn't the only known proof purpose).
5. Finally, return whether the proof is verified or not.

I'm writing this down so it can be turned into a separate issue to further iterate on the algorithm, it doesn't have to go into this PR if that's not desirable at this time.

Reply:
Raised an issue here: https://github.com/w3c/vc-data-integrity/issues/66